File: wrappers/julia/src/rank_cache.jl
"""
    cache_dir() -> String

Return the directory used for caching rank files.
Uses TURBOTOKEN_CACHE_DIR env var or defaults to ~/.cache/turbotoken/.
"""
function cache_dir()::String
    dir = get(ENV, "TURBOTOKEN_CACHE_DIR", "")
    if isempty(dir)
        dir = joinpath(homedir(), ".cache", "turbotoken")
    end
    return dir
end

"""
    ensure_rank_file(name::AbstractString) -> String

Ensure the rank file for the given encoding is downloaded and cached.
Returns the path to the cached file.
"""
function ensure_rank_file(name::AbstractString)::String
    spec = get_encoding_spec(name)
    dir = cache_dir()
    mkpath(dir)
    filepath = joinpath(dir, "$(spec.name).tiktoken")

    if !isfile(filepath)
        @info "Downloading rank file for $(spec.name)..."
        Downloads = Base.require(Base.PkgId(Base.UUID("f43a241f-c20a-4ad4-852c-f6b1247861c6"), "Downloads"))
        Downloads.download(spec.rank_file_url, filepath)
    end

    return filepath
end

"""
    read_rank_file(name::AbstractString) -> Vector{UInt8}

Read the rank file bytes for the given encoding, downloading if necessary.
"""
function read_rank_file(name::AbstractString)::Vector{UInt8}
    filepath = ensure_rank_file(name)
    return read(filepath)
end

File: wrappers/r/R/rank_cache.R
#' Get the cache directory for rank files
#'
#' Uses TURBOTOKEN_CACHE_DIR env var or defaults to ~/.cache/turbotoken/.
#'
#' @return Character string, path to cache directory.
#' @keywords internal
cache_dir <- function() {
  dir <- Sys.getenv("TURBOTOKEN_CACHE_DIR", unset = "")
  if (nchar(dir) == 0) {
    dir <- file.path(path.expand("~"), ".cache", "turbotoken")
  }
  dir
}

#' Ensure a rank file is downloaded and cached
#'
#' @param name Character string, encoding name.
#' @return Character string, path to the cached rank file.
#' @keywords internal
ensure_rank_file <- function(name) {
  spec <- get_encoding_spec(name)
  dir <- cache_dir()
  if (!dir.exists(dir)) {
    dir.create(dir, recursive = TRUE)
  }
  filepath <- file.path(dir, paste0(spec$name, ".tiktoken"))

  if (!file.exists(filepath)) {
    message(sprintf("Downloading rank file for %s...", spec$name))
    utils::download.file(spec$rank_file_url, filepath, mode = "wb", quiet = TRUE)
  }

  filepath
}

#' Read rank file bytes
#'
#' @param name Character string, encoding name.
#' @return Raw vector of rank file bytes.
#' @keywords internal
read_rank_file <- function(name) {
  filepath <- ensure_rank_file(name)
  readBin(filepath, "raw", file.info(filepath)$size)
}
